src/Middleware/UriReplacementHandler.cs

```diff
@@ -0,0 +1,125 @@
+/// </summary>
+public class UriReplacementHandlerOption : IUriReplacementHandlerOption
+{
+    private readonly bool isEnabled;
+
+    private readonly IEnumerable<KeyValuePair<string, string>> replacementPairs;
+
+    /// <summary>
+    /// Creates a new instance of UriReplacementOption.
+    /// </summary>
+    /// <param name="isEnabled">Whether replacement is enabled.</param>
+    /// <param name="replacementPairs">Replacements with the key being a string to match against and the value being the replacement.</param>
+    public UriReplacementHandlerOption(bool isEnabled, IEnumerable<KeyValuePair<string, string>> replacementPairs)
+    {
+        this.isEnabled = isEnabled;
+        this.replacementPairs = replacementPairs;
+    }
+
+    /// <inheritdoc/>
+    public bool IsEnabled()
+    {
+        return isEnabled;
+    }
+
+    /// <inheritdoc/>
+    public Uri? Replace(Uri? original)
+    {
+        if(original is null) return null;
+
+        if(!IsEnabled())
+        {
+            return original;
+        }
+
+        var newUrl = new UriBuilder(original);
+        foreach(var pair in replacementPairs)
+        {
+            newUrl.Path = newUrl.Path.Replace(pair.Key, pair.Value);
+        }
+
+        return newUrl.Uri;
+    }
+}
+
+/// <summary>
+/// Replaces a portion of the URL.
+/// </summary>
+/// <typeparam name="TUriReplacementHandlerOption">A type with the rules used to perform a URI replacement.</typeparam>
+public class UriReplacementHandler<TUriReplacementHandlerOption> : DelegatingHandler where TUriReplacementHandlerOption : IUriReplacementHandlerOption, IRequestOption
+{
+    private readonly TUriReplacementHandlerOption uriReplacement;
+
+    /// <summary>
+    /// Creates a new UriReplacementHandler.
+    /// </summary>
+    /// <param name="uriReplacement">An object with the URI replacement rules.</param>
+    public UriReplacementHandler(TUriReplacementHandlerOption uriReplacement)
+    {
+        this.uriReplacement = uriReplacement;
+    }
+
+    /// <inheritdoc/>
+    protected override async Task<HttpResponseMessage> SendAsync(
+        HttpRequestMessage request, System.Threading.CancellationToken cancellationToken)
+    {
+        ActivitySource? activitySource;
+        Activity? activity;
+        if(request.GetRequestOption<ObservabilityOptions>() is ObservabilityOptions obsOptions)
+        {
+            activitySource = new ActivitySource(obsOptions.TracerInstrumentationName);
+            activity = activitySource.StartActivity($"{nameof(UriReplacementHandler<TUriReplacementHandlerOption>)}_{nameof(SendAsync)}");
+            activity?.SetTag("com.microsoft.kiota.handler.uri_replacement.enable", uriReplacement.IsEnabled());
+        }
+        else
+        {
+            activity = null;
+            activitySource = null;
+        }
+
+        try
+        {
+            var uriReplacementHandlerOption = request.GetRequestOption<TUriReplacementHandlerOption>() ?? uriReplacement;
```

Comment:
There's a potential lookup failure here in the event the user initializes the handler as `new UriReplacementHandler<IUriReplacementHandlerOption>` and then creates a request as 
```cs
var requestInfo = new RequestInformation
{
    HttpMethod = Method.GET,
    URI = new Uri("http://localhost")
};
var uriReplacementHandlerOption = new UriReplacementHandlerOption(false, new Dictionary<string, string> { { "test", "" }}); ;
requestInfo.AddRequestOptions(new IRequestOption[] { uriReplacementHandlerOption });
```

Since the type in the handler is an interface and the requuestOption is a concrete type, the lookup will fail and end up using the wrong option.